File: mobile-v26/app/src/main/java/com/wonrak/emptybottle/StabilityTracker.kt
package com.wonrak.emptybottle

class StabilityTracker(private val requiredHits: Int = 3) {
    private data class Vote(var value: String = "", var hits: Int = 0)
    private val votes = mutableMapOf<String, Vote>()

    fun reset() = votes.clear()

    fun stable(key: String, value: String?): String? {
        if (value.isNullOrBlank()) return null
        val normalized = value.trim().replace(Regex("\\s+"), " ")
        val vote = votes.getOrPut(key) { Vote() }
        if (vote.value == normalized) vote.hits++ else {
            vote.value = normalized
            vote.hits = 1
        }
        return if (vote.hits >= requiredHits) vote.value else null
    }
}

File: mobile_flutter_v26/lib/stability_tracker.dart
class StabilityTracker {
  final int requiredHits;
  final Map<String, _Vote> _votes = {};

  StabilityTracker({this.requiredHits = 3});

  void reset() => _votes.clear();

  String? stable(String key, String? value) {
    if (value == null || value.trim().isEmpty) return null;
    final normalized = value.trim().replaceAll(RegExp(r'\s+'), ' ');
    final vote = _votes.putIfAbsent(key, () => _Vote());

    if (vote.value == normalized) {
      vote.hits++;
    } else {
      vote.value = normalized;
      vote.hits = 1;
    }

    return vote.hits >= requiredHits ? vote.value : null;
  }
}

class _Vote {
  String value = '';
  int hits = 0;
}
